### ISSUE TITLE
Expose unified util to transpile theme components

### ISSUE THREAD

**Issue by Josh-Cena** [facebook/docusaurus#5564]
## 💥 Proposal

The TypeScript theme components are transpiled twice, once only with the types stripped and once to actual commonjs code. Imagine the following case:

- The developer of a plugin writes the plugin in TS;
- Unaware of the double-transpile workflow, the plugin is only built with `tsc` and distributed;
- A user uses the plugin, and decided to swizzle a theme component, but uses the JS version;
- The user will receive an unreadable component.

It is very rare as of now, but can be likely in the future with a better plugin ecosystem. Also, many official plugins have their own theme components, and most of them are kept as JS to bypass this issue—still not optimal for TS-perfectionists like me. If we can expose a util to transpile the theme components to human-readable JS automatically, things would be a lot easier for the plugin authors. Maybe a `docusaurus pack` command? 

### Have you read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes

**Comment by Josh-Cena** [comment]
Made a proof-of-concept package: https://www.npmjs.com/package/@joshcena/docusaurus-plugin-utils

**Comment by slorber** [comment]
Yes I totally agree with this idea and would be happy to have such util in core and use it internally (before exposing it to plugin authors a bit later).

The `copyUntypedFiles` has always been a pain for me + watch mode for css files does not even work everywhere currently (like in theme-common), which is really painful for some devs

**Comment by Josh-Cena** [comment]
@slorber Turns out that this is not that easy and probably requires playing with our infrastructure. If you look at my implementation: https://github.com/jc-verse/docusaurus-plugin-utils there are several things I'm not happy with:

- The `build` function calls an internal API `@babel/cli/lib/babel/dir` which will probably break in future versions. (See https://github.com/babel/babel/issues/9224 maybe)
- To build the other plugins this package has to be built first in our monorepo, which I don't know how to ensure—it seems Lerna builds packages in no particular order, and I can't find where it's configured. I've tested my util in multiple Docusaurus projects and they seem to be working well, but I'm not sure about how it works in a monorepo.
  - Of course, I want to write this util in TypeScript, because JavaScript is whacky.
- We run prettier after this, but Prettier has no API to format a directory either. We would still need to walk the directory ourselves.

**Comment by slorber** [comment]
> The build function calls an internal API @babel/cli/lib/babel/dir which will probably break in future versions.

If we don't expose this package publicly this shouldn't be a problem for internal usage, until we find a solution to provide to the community.

At first I'd prefer to keep this package undocumented until we are sure it works well for us

> To build the other plugins this package has to be built first in our monorepo, which I don't know how to ensure—it seems Lerna builds packages in no particular order, and I can't find where it's configured. I've tested my util in multiple Docusaurus projects and they seem to be working well, but I'm not sure about how it works in a monorepo.

Lerna will build according to the dependency graph. If the new package is listed in theme deps it will be build first. 

That's also why Lerna doesn't allow any dependency cycle otherwise it can't compute a build order.

> Of course, I want to write this util in TypeScript, because JavaScript is whacky.

For such package using JS doesn't feel like a big deal IMHO. It's better to have it in JS than do not have it 🤪 

> We run prettier after this, but Prettier has no API to format a directory either. We would still need to walk the directory ourselves.

glob dir/** + forEach + format should work fine

**Comment by armano2** [comment]
in teory we could use [overrides] for now, and bundle everything with babel in one go, but currently there is no way to apply this to every package, as there is no clear spit between server and client. i read somewhere proposal (maybe comment) about refactoring structure of all packages to:
- src/client
- src/server
but i don't know if that's something what you are planning to do.

Most packages currently follow rule: if theme than this is client, but there are exceptions to that, for example https://github.com/facebook/docusaurus/blob/main/packages/docusaurus-plugin-google-gtag/src/gtag.ts


[overrides]: https://babeljs.io/docs/en/options#overrides

**Comment by slorber** [comment]
Agree we should move into that direction and have client/server/common in each plugin. 
I've also used eslint restricted import in a few apps to avoid unwanted imports/dependencies, didn't know about this boundaries plugin
